Use In-kernel File Copy for Overlayfs and VFS on Linux

sargun opened the pull request:
**- What I did**
Broke out the copy code from overlayfs into its own module, and added support for in-kernel copying of data, as opposed to having to pass all data back through userspace.

Eventually, I'd like to use this copy code in the VFS graphdriver code, but I'd like to work out any worries here before changing the VFS driver.

My end-game, over perhaps 5 different PRs is:
1) Use copy_file_range for copy code, and modularize it so it can be reused [This PR]
2) Switch VFS to use the shared copy code for Linux systems
3) Speed up copy code a bit by parallelizing metadata and copy operations
4) Switch to the quotactl API for project quotas for XFS and EXT4
5) Add project quota support to the VFS driver

Then, I can use XFS / EXT4 with the VFS driver, and be done with life. Not mucking with the VFS and mount stacking code every time a container spins up, or spins down would be great. I tested a slightly more optimized version of the copy that I posted in the PR, and it runs on the ubuntu container with modern reflinking XFS in < 600 ms on ubuntu:latest. For my test, I simply commented out the xattr and mtime-related code, which turns out to have a significant speed up. See step (3) -- we should be able to get similar performance benefits by performing listxattr in parallel. Otherwise, we're paying a 200-1000 microsecond cost for listxattr, which most of the time will be a no-op, while we could be spending that precious time doing real work.

I’d also like to use this as a bouncing-off step for per-container user namespaces.

**- How I did it**
It now checks the kernel version, and uses that to determine if it can use the `copy_file_range` (introduced in Kernel 4.5). If not, it falls back to what it did previously, and copies the data through Go. It limits each copy to a built-in kernel constant of MAX_RW_COUNT (http://elixir.free-electrons.com/linux/latest/source/include/linux/fs.h#L2138). This also prevents us from accidentally overflowing on 32-bit platforms since the constant is smaller than 2GB.

**- How to verify it**
There are already unit tests in the overlay that leverage the copy module. The copy module also has an associated test to ensure that the logic to verify that the copy is using copy_file_range on newer kernels, and it is not on older kernels. 

**- Description for the changelog**
Use in-kernel file copy helper for overlayfs layers on new Kernels

sargun committed:
Separate daemon/graphdriver/overlay/copy into its own package

Signed-off-by: Sargun Dhillon <sargun@sargun.me>

ijc commented on daemon/graphdriver/copy/copy.go:
```diff
@@ -9,32 +9,94 @@ import (
+	if err != nil {
+		return false
+	}
+	if kernel.CompareKernelVersion(*v, kernel.VersionInfo{Kernel: 4, Major: 5, Minor: 0}) < 0 {
```
In general one where possible should not compare kernel versions, you should just try and use the new feature and if you get `ENOSYS` (or whatever the appropriate errno is for the feature in question) apply the fallback from then on. Otherwise you miss out on backports of fixes, either via kernel.org LTS kernels or (more likely) via distros doing backports.

I see there are a couple of uses of `kernel.CompareKernelVersion` in the overlay code already but they both have an override which cases them to only produce warnings, perhaps those are for more complex fallbacks which perhaps aren't easily detectable via errno.

It actually looks like you have the majority of the fallback logic in place and are simply finding the starting point here -- it looks almost like you could just remove this check and assume available until a error code tells you otherwise (I expect there'd be some other tweaks but you are 90% of the way there I think). From something like this I don't think I'd even worry about issuing a warning...

sargun replied:
Do you think it's okay to use in the test to verify the behaviour of the "degradation"?

ijc replied:
I'm not a maintainer here, but if it were me I would simply have the tests force the copyFileRangeDisabled variable to test the degenerate case.

ijc commented on daemon/graphdriver/copy/copy.go:
```diff
@@ -9,32 +9,94 @@ import (
+		return false
+	}
+	if kernel.CompareKernelVersion(*v, kernel.VersionInfo{Kernel: 4, Major: 5, Minor: 0}) < 0 {
+		atomic.StoreInt32(&copyFileRangeDisabled, 1)
```
Atomic ops can be fairly expensive on many platforms (cross cpu sync/locks), is full atomicity strictly required here? It doesn't seem like the end of the world if a few parallel copy attempts try the fast path and hit the fallback and make multiple assignments to a regular `bool` before it becomes globally visible (unless the go memory model is far more relaxed than I expect).

Also I think you would have to use `atomic.LoadInt32` in the check in order to get the benefits of using this atomic store, otherwise the read is not ordered wrt the store. Witht that then you would be doing an atomic read on every copy which adds to the overhead even more.

sargun replied:
I'm fairly sure there's no guarantees about single (simultaneous) goroutine guarantees about access to the graphdriver. Fairly sure that not using atomic here will result in violating the Go memory model. Compared to all the syscalls / IOPs here, I think the atomic ops are cheap here.

cpuguy83 replied:
If this is set 1 time during startup, there is no need for atomic.

cyphar replied:
@cpuguy83 Even better, just do it in `init()` (see https://github.com/moby/moby/pull/34670#discussion_r136038882). Also, I'm a bit confused why this has been moved to a separate package -- does any other graphdriver need this?

cpuguy83 replied:
I don't think so. `vfs` could use it, but it currently uses `archive.CopyWithTar`

sargun replied:
1) I'm intending on using this for VFS eventually, so I split it out.
2) Testing it when it's an independent package is easier.

cyphar commented on daemon/graphdriver/copy/copy.go:
```diff
@@ -9,32 +9,94 @@ import (
 )
 
-func copyRegular(srcPath, dstPath string, mode os.FileMode) error {
+func isCopyFileRangePosssible(srcDir, dstDir string) bool {
```
I'm not sure why you're doing this check for two random directories. I would recommend doing this check *once* inside an `init()` function, and then just checking the global bool directly when deciding between the "legacy copy" and the "new copy".

sargun commented:
I made some changes:
1) It now checks in init() if the syscall is available by checking if it returns ENOSYS, and sets a flag
2) It no longer tries to check for during the course of an entire copy if the srcDir and dstDir are on different filesystems. It now falls back every time it detects this condition.

sargun committed:
Add zero-copy support to copy module

This changeset allows Docker's VFS, and Overlay to take advantage of
Linux's zerocopy APIs.

The copy function first tries to use the ficlone ioctl. Reason being:
 - they do not allow partial success (aka short writes)
 - clones are expected to be a fast metadata operation
See: http://oss.sgi.com/archives/xfs/2015-12/msg00356.html

If the clone fails, we fall back to copy_file_range, which internally
may fall back to splice, which has an upper limit on the size
of copy it can perform. Given that, we have to loop until the copy
is done.

For a given dirCopy operation, if the clone fails, we will not try
it again during any other file copy. Same is true with copy_file_range.

If all else fails, we fall back to traditional copy.

Signed-off-by: Sargun Dhillon <sargun@sargun.me>

sargun commented:
I just rebased on upstream master, and I'm getting interfacer errors after that linter was added. Any idea what's causing that, along with the timeout? None of these errors are occurring locally.

dmcgowan commented on daemon/graphdriver/copy/copy.go:
```diff
@@ -15,26 +15,89 @@ import (
 	defer dstFile.Close()
 
-	_, err = pools.Copy(dstFile, srcFile)
+	_, _, err = unix.Syscall(unix.SYS_IOCTL, dstFile.Fd(), ficlone, srcFile.Fd())
```
We don't need to attempt to do this clone first and should just rely on doing the copy file range syscall. My impression is the number of kernel/fileystem combos which will support this ioctl and do not have support for copy file range is small and do not justify making this attempt on every copy.

sargun replied:
You're right, every file system that supports FICLONE also supports copy_file_range by reflink.

1. The performance for large files of FICLONE seems to be better
2. clones are expected to be a fast metadata operation

What do you think about moving the FICLONE to only run iif every previous FICLONE was successful, and copyFileRangeEnabled is true?  There would be a nicety to this in that we could also keep track of exdev, and stop optimizing when we get our first exdev?

dmcgowan replied:
> The performance for large files of FICLONE seems to be better

Are there benchmarks that show this? Which operations are done by copy file range which increase the overhead over clone. 

> What do you think about moving the FICLONE to only run iif every previous FICLONE was successful, and copyFileRangeEnabled is true?

I really think this is overcomplicating the code. Why not just attempt the copy file range call, and if it gets `== unix.ENOSYS` move on to legacy copy.  We should be doing the same thing we are doing in containerd here, see https://github.com/containerd/containerd/blob/master/fs/copy_linux.go#L34. Once we update to the latest version of containerd this code should merge with it. Based on the previous comment, the containerd should be updated to ensure files larger than 2GB are handled without error.

sargun replied:
Apparently Go handles this okay, even in a 32-bit environment:
```
package main

import (
    "fmt"
    "golang.org/x/sys/unix"
    "os"
)

func main() {
    f1, _ := os.Open("f1")
    f1Info, _ := os.Stat("f1")
    fmt.Println("F1 Size: ", f1Info.Size())

    f2, _ := os.OpenFile("f2", os.O_RDWR|os.O_TRUNC|os.O_CREATE, 0777)
    amountToCopy := f1Info.Size()
	fmt.Println("X: ", int(amountToCopy))
	fmt.Println(int(amountToCopy) == -1073741823)
    for amountToCopy > 0 {
        n, _ := unix.CopyFileRange(int(f1.Fd()), nil, int(f2.Fd()), nil, int(amountToCopy), 0)
        amountToCopy = amountToCopy - int64(n)
    }
}

```
Output:
```
root@titusagent-devvpc7-m4:/# strace -e trace=285 ./main 
F1 Size:  3221225473
X:  -1073741823
true
syscall_377(0x3, 0, 0x4, 0, 0xc0000001, 0) = 0x7ffff000
syscall_377(0x3, 0, 0x4, 0, 0x40001001, 0) = 0x40001001
+++ exited with 0 +++
```

amir73il replied:
Current code looks good to me. 
@dmcgowan, there is no difference with relevant file systems between clone_file_range() and copy_file_range() implementation. The only difference is the MAX_RW_COUNT limit and return value of copied bytes from copy_file_range vs. return value 0 from clone_file_range(), hence the line ret = len; only after clone_file_range().
FYI, that copy_file_range() syscall tries to call clone_file_range() internally is a recent addition to kernel, far beyond v4.5 and the only way to grammatically test it is to try copy_file_range() with more than MAX_RW_COUNT bytes.

dmcgowan replied:
@amir73il thanks for that explanation, I would not have been able to understand that background on my own. Sounds like you are advocating then for keeping the clone attempt because only the most recent kernel's `copy_file_range` call is as fast as the clone call, due to the recent update to have it `clone_file_range` internally? This sounds reasonable to me to keep as is in that case.

amir73il replied:
I am saying current code is fine, because it won't try to clone more than once anyway. Its not that clone_file_range is that much faster than copy_file_range on xfs/btrfs. They are both the same fast metadata operations, only copy_file_range will limit metadata operation to 32bit range, so copying a very large file will require a few more syscalls. probably not a very common case for docker to deal with.

dmcgowan commented on daemon/graphdriver/copy/copy.go:
```diff
@@ -15,26 +15,89 @@ import (
+}
+
+// The maximum amount of bytes Linux will splice
+const maxRwCount = 0x7ffff000
```
Where is this documented? Copy file range takes in `size_t` which can be 64-bit http://man7.org/linux/man-pages/man2/copy_file_range.2.html.

sargun replied:
See: 
http://elixir.free-electrons.com/linux/v4.12.10/source/fs/read_write.c#L1621
http://elixir.free-electrons.com/linux/v4.12.10/source/include/linux/fs.h#L2138
`INT_MAX` is pretty much always based on 4-byte ints on Linux, on our currently supported platforms, and have 4096 byte pages.
 
For the majority of filesystems underneath us, we're going to be falling back to splice.

dmcgowan replied:
In that case, we still don't need to define this variable here. We should just detect when less than the whole file was written and attempt the copy multiple times. However we can still pass in the full length and when clone is done there is no need to make multiple calls here.

Is this missing setting `ret = len;` here
http://elixir.free-electrons.com/linux/v4.12.10/source/fs/read_write.c#L1617

sargun commented:
Made a few changes:
1) No longer do we chunk up the copy_file_range. See above comments.
2) We do not try to detect if copy_file_range is unavailable at start time, only per copy. There is literally no reason to waste time doing these syscalls on every subsequent file, if we can avoid it. Again, the eventual idea is to hoist this, so VFS can use it.

I did not remove the clone. See this thread: http://oss.sgi.com/archives/xfs/2015-12/msg00356.html

In fact, when @amir73il was working on his patchset for in-kernel ovl copy_up, a similar approach was suggested: https://patchwork.kernel.org/patch/9321675/

amir73il commented on daemon/graphdriver/copy/copy.go:
```diff
@@ -24,20 +29,61 @@ const (
+		}
+
+		*copyWithFileClone = false
+		if err == unix.EXDEV {
```
I think you meant to check EOPNOTSUPP? not sure why EXDEV is relevant here at all. is any of the storage drivers using this expected to copy across mount boundaries??

amir73il commented on daemon/graphdriver/copy/copy.go:
```diff
@@ -24,20 +29,61 @@ const (
+	if *copyWithFileRange {
+		err = doCopyWithFileRange(srcFile, dstFile, fileinfo)
+		// Trying the file_clone may not have caught the exdev case
+		// as the ioctl may not have been available (therefore EINVAL)
```
did you mean (therfore ENOSYS)? again, not sure which case the EXDEV is trying to handle. aren't all copies under /var/lib/docker/<driver> which is supposed to be under a single mount?

sargun replied:
This checks for both exdev and enosys. If FICLONE failed because the ioctl wasn't available, then it falls into this case. I think that checking for EXDEV as well to disable subsequent fast copies is a simple enough check to future-proof us, where it's worth it. 

I can definitely see an admin moving layers "by hand" to another device to spread space usage.

amir73il replied:
Got you. I misread the comment. Nothing wrong with the code. I was just confused because you mentioned the case of kernel with no clone ioctl, but not the common case of fs with no clone support (e.g. ext4) which would return EOPNOTSUPP

sargun replied:
So, good to go?

amir73il replied:
LGTM

amir73il reviewed (changes requested):
Nice work! a few minor corrections

sargun commented:
@dmcgowan, nudge?

AkihiroSuda commented:
ping @dmcgowan

AkihiroSuda commented on daemon/graphdriver/copy/copy.go:
```diff
@@ -15,26 +15,75 @@ import (
-type copyFlags int
+// FICLONE ioctl number -- defined in <linux/fs.h> -- We don't include it
+// because it'd break builds with older versions of kernel headers
+const ficlone = 0x40049409
```
Isn't this arch-specific?

sargun replied:
It is (ish). It applies to all of the architectures other than:
* Power
* Alpha
* MIPS
* Parisc
* Sparc

I can fix this, if you think these architectures should be supported.

AkihiroSuda replied:
At least ppc64le needs to be supported.
For other architectures, please disable the added code?

sargun replied:
Okay, I can actually fix it by adding the IOW in ifndef.

thaJeztah commented:
ping @sargun (https://github.com/moby/moby/pull/34670#discussion_r139652124) @dmcgowan PTAL

sargun commented:
How do I work around the failure for interfacer?
```

17:30:43 daemon/graphdriver/copy/copy.go:89:17:warning: srcFile can be io.Reader (interfacer)
17:30:43 daemon/graphdriver/copy/copy.go:89:26:warning: dstFile can be io.Writer (interfacer)

```

dmcgowan commented:
The arguments to `legacyCopy` should be `legacyCopy(dst io.Writer, src io.Reader)`, the interfacer is enforcing that the argument types have the slimmest interface based on how they are used in the function.

sargun commented:
Good to go?

cpuguy83 commented:
Discussed on the maintainers meeting and it seems like we are good on the design (@dmcgown please mention if you think there's still something to address), so moving this to code review.

sargun commented:
*gentle nudge*

cpuguy83 reviewed (approved):
LGTM

dmcgowan reviewed (approved):
LGTM

sargun commented:
Rebasing.

thaJeztah commented:
All green now, merging. Thanks so much @sargun!
